src/compiler/checker.ts

```diff
@@ -21218,19 +21218,31 @@ namespace ts {
             }
             return literalTypesWithSameBaseType(types) ?
                 getUnionType(types) :
-                reduceLeft(types, (s, t) => isTypeSubtypeOf(s, t) ? t : s)!;
+                reduceLeft(types, (s, t) => isTypeSubtypeOf(getNonNullableTypeIfNeeded(s), getNonNullableTypeIfNeeded(t)) ? t : s)!;
         }
 
         function getCommonSupertype(types: Type[]): Type {
             if (!strictNullChecks) {
                 return getSupertypeOrUnion(types);
             }
-            const primaryTypes = filter(types, t => !(t.flags & TypeFlags.Nullable));
-            if (primaryTypes.length) {
-                const supertypeOrUnion = getSupertypeOrUnion(primaryTypes);
-                return primaryTypes === types ? supertypeOrUnion : getUnionType([supertypeOrUnion, ...filter(types, t => !!(t.flags & TypeFlags.Nullable))]);
+
+            const supertypeOrUnion = getSupertypeOrUnion(types);
+
+            const shouldHaveNull = some(types, t => someHas(t, TypeFlags.Null));
+            const shouldHaveUndefined = some(types, t => someHas(t, TypeFlags.Undefined));
+
+            let missingNullableFlags: TypeFlags = 0;
+            if (shouldHaveNull && !someHas(supertypeOrUnion, TypeFlags.Null)) {
+                missingNullableFlags |= TypeFlags.Null;
+            }
+            if (shouldHaveUndefined && !someHas(supertypeOrUnion, TypeFlags.Undefined)) {
+                missingNullableFlags |= TypeFlags.Undefined;
+            }
+            return getNullableType(supertypeOrUnion, missingNullableFlags);
+
+            function someHas(type: Type, flags: TypeFlags): boolean {
+                return someType(type, t => !!(t.flags & flags));

```

Comment:
This feels common enough that I would rather us lift it up to `someTypeHasFlags`

Reply:
I can do that, yeah, though I think most places just entirely inline it, so I could just do that too.